middle_end/flambda2/from_lambda/lambda_to_lambda_transforms.ml

```diff
@@ -33,6 +33,53 @@ let switch_for_if_then_else ~cond ~ifso ~ifnot ~kind =
   in
   L.Lswitch (cond, switch, L.try_to_find_location cond, kind)
 
+(* This function helps bind expression to avoid duplicating them in the
+   generated code. Currently, this is only used for if-then-else optimization,
+   which guarantees that even if expressions are duplicated, they are still only
+   evaluated once (but could have been duplicated among multiple code paths), so
+   this is for code size optimization mainly.
+
+   One could want to duplicate constants, however the backend does not
+   deduplicate them, especially when they are the return value of the function,
+   so for now constants returned are shared via a static catch/raise. *)
+let bind_expr ~kind ~expr k =
+  let is_simple_duplicable expr =
+    match[@warning "-4"] (expr : L.lambda) with
+    | Lvar _ | Lconst _ -> true
+    | _ -> false
+  in
+  match[@warning "-4"] (expr : L.lambda) with
+  | Lstaticraise (_, args) when List.for_all is_simple_duplicable args -> k expr
+  | _ ->
+    let cont = L.next_raise_count () in
+    let jump = L.Lstaticraise (cont, []) in
+    L.Lstaticcatch (k jump, (cont, []), expr, Same_region, kind)
+
+let optimize_boolean_sequop ~cond ~ifso ~ifnot ~kind =
+  let rec aux ~kind ~cond ~ifso ~ifnot =
+    match[@warning "-4"] cond with
+    (* CR gbury: should we try to use the locs here, or is it better to keep
+       using the locs from each individual condition ? *)
```

Comment:
`mk_switch` contains a call to `L.try_to_find_location`. I suspect that we may want to instead optionally pass a location, only using `try_to_find_loaction` if we don't have one. But I'm fine with this PR keeping the CR, and coming back to this later.